### PR TITLE
Add Include what you use

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Use LLVM5 and IWYU 0.9 with Travis
- Add mapping files for IWYU (see https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUMappings.md)
- Run IWYU over entire source code and fix issues
- Make the `tests/Unit/TestingFramework.hpp` be the first include in all test source files. This resolves the ADL troubles we occasionally run into with some source files.

I'll add documentation on how to use the tools we supply in a dev guide entry later.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [x] Build system
- [x] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

File: tools/Iwyu/iwyu.imp [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+[
+  { symbol: ["size_t", "private", "<cstddef>", "public"]},
+  { symbol: ["gr::Tags::InverseSpacetimeMetric", private,
```
Comment (kidder) [inline tools/Iwyu/iwyu.imp]: remove the two tags

File: tools/Iwyu/iwyu.imp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  { include: ["<catch.hpp>", "public",
+              "\"tests/Unit/TestingFramework.hpp\"", "public"]},
+
+  { include: ["<type_traits>", "public",
+              "\"Utilities/TypeTraits.hpp\"", "public"]},
+
+  { include: ["@<H5.*public.h>", "private",
+              "<hdf5.h>", "public"]},
+
+  { include: ["<numpy/__multiarray_api.h>", private,
```
Comment (kidder) [inline tools/Iwyu/iwyu.imp]: use consistent quoting for private and public

File: .travis/RunIncludeWhatYouUse.sh [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    then
+        printf "\nIWYU found problems!\n\
+ #     # ####### ####### #######  #\n\
+ ##    # #     #    #    #       ###\n\
+ # #   # #     #    #    #        #\n\
+ #  #  # #     #    #    #####\n\
+ #   # # #     #    #    #        #\n\
+ #    ## #     #    #    #       ###\n\
+ #     # #######    #    #######  #\n\
+IWYU is still not yet fully mature and  provides incorrect suggests\n\
```
Comment (kidder) [inline .travis/RunIncludeWhatYouUse.sh]: suggestions

File: .travis/RunIncludeWhatYouUse.sh [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  various dimensions, say up to Dim=4, quite easily in a cpp file. Thus, the\n\
+  header only has a forward declaration and the function definition and\n\
+  instantiations are in the cpp file. IWYU will tell you that you do not\n\
+  need to include DataStructures/Index.hpp in the cpp file, which is\n\
+  incorrect. You can tell IWYU that it is wrong and you must include the\n\
+  header file by including it as:\n
+  #include \"DataStructures/Index.hpp\"  // IWYU pragma: keep\n\
+- Class template declarations with default template parameters are not\n\
+  handled correctly. See issue #530 on IWYU GitHub. Specifically, say you\n\
+  Have a header TagsDeclarations.hpp that declares\n\
```
Comment (kidder) [inline .travis/RunIncludeWhatYouUse.sh]: have

File: .travis/RunIncludeWhatYouUse.sh [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  'template <size_t Dim, class Frame = Frame::Inertial> MyTag;' and another\n\
+  header file Tags.hpp that defines the class. Then it suffices and is\n\
+  optimal to include TagsDeclarations.hpp in other header files that do not\n\
+  need the definition of the Tags. However, IWYU will tell you to provide the\n\
+  forward declaration with the defaulted template parameters in the header\n\
+  file that you'd include TagsDeclarations.hpp in. This means you are\n\
+  the default template parameters in more than one place, which is not\n\
```
Comment (kidder) [inline .travis/RunIncludeWhatYouUse.sh]: missing "including"

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Comment by kidder** [comment]
tests do not pass because of timeout of clang-tidy since every file was touched

**Pull request merged**